code/xcvm/evm/src/Gateway.sol

```diff
@@ -0,0 +1,145 @@
+    mapping(uint256 => mapping(BridgeSecurity => address)) public bridges;
+    mapping(uint256 => address) public assets;
+
+    event InstanceCreated(uint256 networkId, bytes user, address instance);
+
+    event Spawn(
+        bytes account,
+        uint256 networkId,
+        BridgeSecurity security,
+        uint256 salt,
+        bytes spawnedProgram,
+        address[] assetAddresses,
+        uint256[] amounts
+    );
+
+    modifier onlyBridge() {
+        require(bridgesInfo[msg.sender].security != BridgeSecurity(0));
+        _;
+    }
+
+    constructor() {
+        // enable trustless bridge;
+    }
+
+    function getAsset(uint256 assetId) external returns (address) {
+        return assets[assetId];
+    }
+
+    function getBridge(uint256 networkId, BridgeSecurity security) external returns (address) {
+        return bridges[networkId][security];
+    }
+
+    function registerAsset(address assetAddress, uint128 assetId) external onlyOwner {
+        require(assetAddress != address(0), "Gateway: invalid address");
+        assets[assetId] = assetAddress;
+    }
+
+    function unregisterAsset(uint128 assetId) external onlyOwner {
+        delete assets[assetId];
+    }
+
+    function registerBridge(
+        address bridgeAddress,
+        BridgeSecurity security,
+        uint256 networkId
+    ) external onlyOwner {
+        require(bridges[networkId][security] == address(0), "Gateway: this type of bridge already registered");
+        require(bridgeAddress != address(0), "Gateway: invalid address");
+        require(bridgesInfo[bridgeAddress].security == BridgeSecurity(0), "Gateway: bridge already enabled");
+        require(security != BridgeSecurity(0), "Gateway: should not disable bridge while registering bridge");
+        bridgesInfo[bridgeAddress].security = security;
+        bridgesInfo[bridgeAddress].networkId = networkId;
+        bridges[networkId][security] = bridgeAddress;
+    }
+
+    function unregisterBridge(address bridgeAddress) external onlyOwner {
+        require(bridgesInfo[bridgeAddress].security != BridgeSecurity(0), "Gateway: bridge already disabled");
+        delete bridges[bridgesInfo[bridgeAddress].networkId][bridgesInfo[bridgeAddress].security];
+        bridgesInfo[bridgeAddress].security = BridgeSecurity(0);
+        bridgesInfo[bridgeAddress].networkId = 0;
+    }
+
+    //// TODO ? is the bridge who's gonna to provide internetwork assets transfer?
+    function _provisionAssets(
+        address payable interpreterAddress,
+        address[] memory erc20AssetList,
+        uint256[] memory amounts
+    ) internal {
+        require(
+            erc20AssetList.length == amounts.length,
+            "Gateway: asset list size shuold be equal to amount list size"
+        );
+        if (msg.value > 0) {
+            bool sent = interpreterAddress.send(msg.value);
+            require(sent, "Failed to send Ether");
+        }
+        for (uint256 i = 0; i < erc20AssetList.length; i++) {
```

Comment:
I guess `++i` is more gas efficient than `i++`. https://www.reddit.com/r/ethdev/comments/tcwspw/i_vs_i_gas_efficiency/

Reply:
And also `uint256 i` is already `0`.